### PR TITLE
Types: Add support for CrateDB's `FLOAT_VECTOR` data type and `KNN_MATCH` function

### PR DESCRIPTION
**Pull request by amotl**
## About

CrateDB's [`FLOAT_VECTOR`](https://crate.io/docs/crate/reference/en/latest/general/ddl/data-types.html#float-vector) data type, and its accompanying [`KNN_MATCH`](https://cratedb.com/docs/crate/reference/en/latest/general/builtins/scalar-functions.html#scalar-knn-match) function, have been added with CrateDB 5.5.0. This patch adds corresponding SQLAlchemy type and function/predicate support.

## Details

Those two implementetations have been predecessors, and both sites will be used to verify this patch, after it has been merged.

- [LangChain Vector Store: vectorstores/cratedb/sqlalchemy_type.py](https://github.com/crate-workbench/langchain/blob/0a831a6/libs/langchain/langchain/vectorstores/cratedb/sqlalchemy_type.py)
- [Meltano CrateDB Target: sqlalchemy/vector.py](https://github.com/crate-workbench/meltano-target-cratedb/blob/5c30efc/target_cratedb/sqlalchemy/vector.py)

## Backlog
- [x] Documentation
- [x] Software tests

### CODE REVIEW

File: src/sqlalchemy_cratedb/type/vector.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+
+    def compile_column(self, compiler):
+        if isinstance(self.column, dict):
+            column = ', '.join(
+                sorted(["{0} {1}".format(compiler.process(k), v)
+                       for k, v in self.column.items()])
+            )
+            return "({0})".format(column)
+        else:
+            return "{0}".format(compiler.process(self.column))
```
[the remark this answers is not in the document]
Comment (amotl) [reply src/sqlalchemy_cratedb/type/vector.py:193]: Thinking about it once more, I think it must stay, in order to make matching on nested elements of ObjectType columns work?

[the remark this answers is not in the document]
Comment (amotl) [reply src/sqlalchemy_cratedb/type/vector.py:193]: Actually trying to verify this case, it looks like `FLOAT_VECTOR` columns need to be dedicated first-level columns. They can't be used within `OBJECT` type values.

```
UnsupportedFeatureException[knn_match can only be used in WHERE clause for tables as it needs an index to operate on]
```

[the remark this answers is not in the document]
Comment (amotl) [reply src/sqlalchemy_cratedb/type/vector.py]: This fragment has been removed now.

```python
if isinstance(self.column, dict):
    column = ', '.join(
        sorted(["{0} {1}".format(compiler.process(k), v)
               for k, v in self.column.items()])
    )
    return "({0})".format(column)
```

File: src/sqlalchemy_cratedb/type/vector.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+          `ARRAY` right away, to get a few of the features without
+          the need to redefine them?
+
+    Please note the outcome of this analysis and the corresponding implementation
+    has been derived from empirical observations, and from the feeling that we also
+    lack corresponding support on the other special data types of CrateDB (ARRAY and
+    OBJECT) within the SQLAlchemy dialect, i.e. "that something must be wrong or
+    incomplete". In this spirit, it is advisable to review and improve their
+    implementations correspondingly.
+    """
```
[the remark this answers is not in the document]
Comment (amotl) [reply src/sqlalchemy_cratedb/type/vector.py:140]: I've diverted these implementation notes to GH-65, and removed them from the code.

File: src/sqlalchemy_cratedb/type/vector.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+class KnnMatch(ColumnElement):
+    """
+    Wrap CrateDB's `KNN_MATCH` function into an SQLAlchemy function.
+
+    https://cratedb.com/docs/crate/reference/en/latest/general/builtins/scalar-functions.html#scalar-knn-match
+    """
+    inherit_cache = True
+
+    def __init__(self, column, term, k=None):
+        super(KnnMatch, self).__init__()
```
Comment (surister) [inline src/sqlalchemy_cratedb/type/vector.py:134]: This looks like python2 super call, why not just `super().__init__()`

Comment (amotl) [reply src/sqlalchemy_cratedb/type/vector.py:134]: Good catch. Probably a copy/paste problem. Thanks.

Comment (amotl) [reply src/sqlalchemy_cratedb/type/vector.py]: Fixed.

**Review by matriv — APPROVED** [review]
thank you!

**Pull request merged**